src/wp-includes/class-wp-modules.php

```diff
@@ -0,0 +1,251 @@
+		if ( ! isset( $this->registered[ $module_identifier ] ) ) {
+			$deps = array();
+			foreach ( $dependencies as $dependency ) {
+				if ( isset( $dependency['id'] ) ) {
+					$deps[] = array(
+						'id'   => $dependency['id'],
+						'type' => isset( $dependency['type'] ) && 'dynamic' === $dependency['type'] ? 'dynamic' : 'static',
+					);
+				} elseif ( is_string( $dependency ) ) {
+					$deps[] = array(
+						'id'   => $dependency,
+						'type' => 'static',
+					);
+				}
+			}
+
+			$this->registered[ $module_identifier ] = array(
+				'src'          => $src,
+				'version'      => $version,
+				'enqueued'     => isset( $this->enqueued_before_registered[ $module_identifier ] ),
+				'dependencies' => $deps,
+			);
+		}
+	}
+
+	/**
+	 * Marks the module to be enqueued in the page.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $module_identifier The identifier of the module.
+	 */
+	public function enqueue( $module_identifier ) {
+		if ( isset( $this->registered[ $module_identifier ] ) ) {
+			$this->registered[ $module_identifier ]['enqueued'] = true;
+		} else {
+			$this->enqueued_before_registered[ $module_identifier ] = true;
+		}
+	}
+
+	/**
+	 * Unmarks the module so it is no longer enqueued in the page.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $module_identifier The identifier of the module.
+	 */
+	public function dequeue( $module_identifier ) {
+		if ( isset( $this->registered[ $module_identifier ] ) ) {
+			$this->registered[ $module_identifier ]['enqueued'] = false;
+		}
+		unset( $this->enqueued_before_registered[ $module_identifier ] );
+	}
+
+	/**
+	 * Returns the import map array.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return array Array with an `imports` key mapping to an array of module identifiers and their respective URLs,
+	 *               including the version query.
+	 */
+	public function get_import_map() {
+		$imports = array();
+		foreach ( $this->get_dependencies( array_keys( $this->get_enqueued() ) ) as $module_identifier => $module ) {
+			$imports[ $module_identifier ] = $module['src'] . $this->get_version_query_string( $module['version'] );
+		}
+		return array( 'imports' => $imports );
+	}
+
+	/**
+	 * Prints the import map using a script tag with a type="importmap" attribute.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_import_map() {
+		$import_map = $this->get_import_map();
+		if ( ! empty( $import_map['imports'] ) ) {
+			wp_print_inline_script_tag(
+				wp_json_encode( $import_map, JSON_HEX_TAG | JSON_HEX_AMP ),
+				array(
+					'type' => 'importmap',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Prints all the enqueued modules using script tags with type="module"
+	 * attributes.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_enqueued_modules() {
+		foreach ( $this->get_enqueued() as $module_identifier => $module ) {
+			wp_print_script_tag(
+				array(
+					'type' => 'module',
+					'src'  => $module['src'] . $this->get_version_query_string( $module['version'] ),
```

Comment:
While not common, it's possible that the `src` here may already have a query string (such as when importing a script from a service on another origin). Therfore, I think `add_query_arg` should rather be used here. The method can be changed to return an array:

```suggestion
					'src'  => add_query_arg( $this->get_version_query_arg( $module['version'] ), $module['src'] ),
```

Reply:
Or even better, there should be a new method introduced like `get_versioned_src( $module )` which returns the string instead of relying on concatenation in 3 places here. So this would be:

```suggestion
					'src'  => $this->get_versioned_src( $module ),
```

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/5818/commits/1df672d3da05ae7bfdf52caffac3a9d4d38752e1.